### PR TITLE
Deduplicate blocks in any hash join

### PR DESCRIPTION
**Pull request by vdimir**
### Changelog category (leave one):

- Performance Improvement

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
* Do not store blocks in `ANY` hash join if nothing is inserted

### Documentation entry for user-facing changes

- [ ] Documentation is written (mandatory for new features)

> Information about CI checks: https://clickhouse.com/docs/en/development/continuous-integration/

### CODE REVIEW

**Comment by Marku555** [comment]
Hello!
Any news on when this PR will be merged to the master?

**Comment by vdimir** [comment]
@Marku555 I'll continue working on PR this week

**Comment by robot-clickhouse-ci-1** [comment]
This is an automated comment for commit 62e96b55a4ec0f083823c4f29456aa89b85762ac with description of existing statuses. It's updated for the latest CI running
The full report is available [here](https://s3.amazonaws.com/clickhouse-test-reports/48633/62e96b55a4ec0f083823c4f29456aa89b85762ac/ci_running.html)
The overall status of the commit is 🟡 pending

<table><thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody><tr><td>CI running</td><td>A meta-check that indicates the running CI. Normally, it's in <b>success</b> or <b>pending</b> state. The failed status indicates some problems with the PR</td><td>🟡 pending</td></tr>
<tr><td>Mergeable Check</td><td>Checks if all other necessary checks are successful</td><td>🟢 success</td></tr>
</table>

**Comment by vdimir** [comment]
> Stress test (ubsan) — Killed by signal (in clickhouse-server.log) [Details](https://s3.amazonaws.com/clickhouse-test-reports/48633/62e96b55a4ec0f083823c4f29456aa89b85762ac/stress_test__ubsan_.html)

Solved https://github.com/ClickHouse/ClickHouse/issues/49102

> Integration tests (tsan) [2/6] — fail: 1, passed: 250, flaky: 0 [Details](https://s3.amazonaws.com/clickhouse-test-reports/48633/62e96b55a4ec0f083823c4f29456aa89b85762ac/integration_tests__tsan__[2_6].html)

`test_merge_tree_s3_restore` https://github.com/ClickHouse/ClickHouse/issues/48498

**Comment by vdimir** [comment]
@devcrafter do you want to review?

File: tests/queries/0_stateless/02725_any_join_single_row.sql [hunk 1]
```diff
@@ -0,0 +1,26 @@
+CREATE TABLE join_test
+(
+    `key` UInt64,
+    `value` UInt64
+)
+ENGINE = Join(ANY, LEFT, key);
```
Comment (devcrafter) [inline tests/queries/0_stateless/02725_any_join_single_row.sql:6]: As discussed, let's add a test for RIGHT join

**Comment by vdimir** [comment]
> Stateless tests (release, DatabaseReplicated) [3/4] — fail: 1, passed: 1263, skipped: 33 [Details](https://s3.amazonaws.com/clickhouse-test-reports/48633/bdb192cf2742d6f3059f621068bbc59d78124229/stateless_tests__release__databasereplicated__[3_4].html)
> Stateless tests (release, s3 storage) [1/2] — fail: 1, passed: 2604, skipped: 28 [Details](https://s3.amazonaws.com/clickhouse-test-reports/48633/bdb192cf2742d6f3059f621068bbc59d78124229/stateless_tests__release__s3_storage__[1_2].html)
> Stateless tests (release, analyzer) — fail: 1, passed: 5304, skipped: 17 [Details](https://s3.amazonaws.com/clickhouse-test-reports/48633/bdb192cf2742d6f3059f621068bbc59d78124229/stateless_tests__release__analyzer_.html)

`01103_check_cpu_instructions_at_startup`



> Stress test (asan) — Killed by signal (in clickhouse-server.log) [Details](https://s3.amazonaws.com/clickhouse-test-reports/48633/bdb192cf2742d6f3059f621068bbc59d78124229/stress_test__asan_.html)
> Stress test (ubsan) — Killed by signal (in clickhouse-server.log) [Details](https://s3.amazonaws.com/clickhouse-test-reports/48633/bdb192cf2742d6f3059f621068bbc59d78124229/stress_test__ubsan_.html)

https://github.com/ClickHouse/ClickHouse/issues/49793


> Upgrade check (tsan) — Possible deadlock on shutdown (see gdb.log) [Details](https://s3.amazonaws.com/clickhouse-test-reports/48633/bdb192cf2742d6f3059f621068bbc59d78124229/upgrade_check__tsan_.html)

https://github.com/ClickHouse/ClickHouse/issues/33531

**Comment by vpvha** [comment]
Hello! 
Wil be this commit backport to 22.8v?

**Comment by Marku555** [comment]
@vdimir 
Hello! We use the 23.3 LTS release in the production and the problem bothers us. Is it possible to make a backport to 23.3?

**Pull request merged**